### PR TITLE
add persistent app config

### LINKED ISSUE #24: persistent state
Currently the app loses the torrent list every time the main window closes or is reloaded, and when the app is restarted. Ideally state is stored on disk somewhere so that sessions are just resuming what's already been in progress beforehand.

Looking for good modules (or just vanilla solutions, like a state.json?) to solve this issue. Currently avoiding anything that requires native dependencies or grows the dependency graph in a big way.

- https://github.com/LinusU/node-application-config
    - This module is used by ghauth and has worked well for me in the past. Might be too small for handling things like the torrent list and video playback state once more things need to be tracked, but works for basics and would still be useful for storing application preferences in the future (like default download directory).

### PR DESCRIPTION
**Pull request by ngoldman**
Using `application-config` for now. So far working great. Storing magnet URIs of torrents in config instead of `.torrent` file paths since those are often deleted.

Played around with some different ways to capture and save torrent info until I realized I could just save info from `state.client.map` with the props I wanted on every client `torrent` event. That way no deduping is necessary. The WebTorrent client always tells the truth.

> resolves #24

### CODE REVIEW

**Comment by ngoldman** [comment]
One important thing to note, this does NOT implement deleting from the config.

**Comment by ngoldman** [comment]
Okay now deleting torrents persists.

**Comment by ngoldman** [comment]
Should be noted, `torrent.destroy` doesn't remove files from `/tmp/webtorrent`. That's probably the desired behavior once we get a default & custom download folder into prefs.

File: renderer/index.js [hunk 1]
```diff
@@ -251,6 +251,35 @@ function addTorrentEvents (torrent) {
...
+    return {
+      name: torrent.name,
+      magnetURI: torrent.magnetURI,
+      path: torrent.path
+    }
+  })
+
+  cfg.write({
+    torrents: state.saved.torrents
+  }, function (err) {
```
Comment (feross) [inline renderer/index.js]: Why not `cfg.write(state.saved)` to save the whole thing?

File: renderer/index.js [hunk 2]
```diff
@@ -286,7 +315,9 @@ function openPlayer (torrent) {
 }
 
 function deleteTorrent (torrent) {
-  torrent.destroy(update)
+  torrent.destroy(function () {
+    saveTorrents() // updates after writing to config
```
Comment (feross) [inline renderer/index.js]: Good call. We can add some indication to the UI that it's being deleted later, so the button feels responsive.

File: renderer/index.js [hunk 3]
```diff
@@ -251,6 +251,35 @@ function addTorrentEvents (torrent) {
   }
 }
 
+function restoreSession () {
+  cfg.read(function (err, data) {
+    if (err) console.error(err)
+    state.saved = data
+    if (!state.saved.torrents) state.saved.torrents = []
```
Comment (feross) [inline renderer/index.js]: Why is this necessary?

**Comment by feross** [comment]
THIS IS OFFICIALLY A REAL APP NOW. :rocket:

File: renderer/index.js (hunk 1 above)
Comment (ngoldman) [inline renderer/index.js]: Thought that might lead to overwrites in the future, but that would be fine for now.

File: renderer/index.js (hunk 3 above)
Comment (ngoldman) [inline renderer/index.js]: On initial read of config it's an empty object, so state.saved.torrents.forEach would throw an error.

Comment (feross) [inline renderer/index.js]: If we detect it's an empty object, can't we just not do `state.saved = data`, since `state.saved.torrents` is already set to  an initial value in state.js?

Comment (ngoldman) [inline renderer/index.js]: Truth.
